[AutoFill Debugging] Add infrastructure for JavaScript-based text extraction filtering rules

whsieh opened the pull request:
#### 9ecfd550499977ec42beb807070ba9d50bef4537
<pre>
[AutoFill Debugging] Add infrastructure for JavaScript-based text extraction filtering rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=303277">https://bugs.webkit.org/show_bug.cgi?id=303277</a>
<a href="https://rdar.apple.com/165583914">rdar://165583914</a>

Reviewed by Abrar Rahman Protyasha.

Add basic infrastructure to plumb and evaluate JavaScript-based text extraction filtering rules in a
separate, internal script world as a part of text extraction filtering. Note that there is no change
in behavior yet, since there&apos;s no logic to fetch the remotely updatable rule list yet.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRules):

Add a helper method to convert a list of `FilterRuleData` to `FilterRule`, by initializing the URL
regex (if specified).

(WebCore::TextExtraction::filteringWorld):

Set up an internal script world for evaluating text extraction filtering rules.

(WebCore::TextExtraction::applyRules):

Add a helper method that takes a list of text extraction rules, and applies the rules to the `input`
text, each of which consists of a JavaScript snippets to evaluate in the above internal script
world, as well as an optional regex to match against the URL of the given frame.

Each JavaScript snippet takes as arguments the original text, and the container node enclosing the
text; it asynchronously resolves to either a filtered version of the text, or `null` (indicating
that filtering is not necessary).

All snippets are evaluated in parallel, and (to err on the side of the most restrictive rule) the
shortest result is surfaced back through the final completion handler.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

Add a couple new structs, representing data used to initialize filtering rules, and the filtering
rules themselves.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):

If the below option flag is set, add another callback to the filtering pipeline that calls out to
the web process to evaluate the extracted text against filtering rules.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:

Add a new option, `_WKTextExtractionFilterRules`, which controls whether or not to enable rule-based
text filtering.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTextExtractionFilterRules):
(WebKit::WebPageProxy::applyTextExtractionFilter):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add plumbing from the UI -&gt; web process (see above).

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateTextExtractionFilterRules):
(WebKit::WebPage::applyTextExtractionFilter):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/303657@main">https://commits.webkit.org/303657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e528fa236941fa61f3b55355d871a4e73745b78

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133213 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140767 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85258 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ee2b60ae-b44f-4394-8fb5-33242e434a99) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135083 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5579 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101891 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69351 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/59599739-0d8e-4ba5-af27-252ea7219832) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4415 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119402 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82686 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f6d92a55-7567-4920-8ad5-964739e44c58) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4300 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1876 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37517 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143415 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5384 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38094 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110268 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5466 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4630 "Found 1 new test failure: imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-root-block-scroll.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110451 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27982 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4170 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115656 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59129 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5439 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34010 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5285 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68891 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5528 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5395 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1773,5 +1784,105 @@ std::optional<SimpleRange> rangeForExtractedText(const LocalFrame& frame, Extrac
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&] {
+        world.get() = DOMWrapperWorld::create(commonVM(), DOMWrapperWorld::Type::Internal, "Text Extraction Filtering Rules"_s);
+    });
```
We have thread-safe statics enabled in WebCore now, so I think you can drop the `std::call_once`? https://github.com/WebKit/WebKit/commit/2d4ca326601eb0a377673f8200349913aae83c65




```suggestion
    static NeverDestroyed<RefPtr<DOMWrapperWorld>> world = DOMWrapperWorld::create(commonVM(), DOMWrapperWorld::Type::Internal, "Text Extraction Filtering Rules"_s);
```

whsieh replied:
Good point — removed the `call_once` here. Thanks for the review!

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1773,5 +1784,105 @@ std::optional<SimpleRange> rangeForExtractedText(const LocalFrame& frame, Extrac
+        return std::make_optional(WTFMove(argumentMap));
+    }();
+
+    auto filteredStrings = Box<Vector<String>>::create();
```
random musing, I wonder if we should start using safer-cpp guidance for `Box` values too, given the refcounted semantics

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1773,5 +1784,105 @@ std::optional<SimpleRange> rangeForExtractedText(const LocalFrame& frame, Extrac
+
+    auto urlString = document->url().string();
+    for (auto& [name, urlPattern, source] : rules) {
+        if (urlPattern && urlPattern->match(urlString) < 0)
```
We also want to skip the iteration when `urlPattern == std::nullopt`, right? something like this?
```suggestion
        if (!urlPattern || urlPattern->match(urlString) < 0)
```

aprotyas replied:
or maybe `urlPattern == std::nullopt` means allow any URL pattern.

whsieh replied:
Ah, that's right — optional is intended to default to "apply regardless". I can make this more explicit by making it a `Variant` instead.

whsieh committed:
[AutoFill Debugging] Add infrastructure for JavaScript-based text extraction filtering rules
https://bugs.webkit.org/show_bug.cgi?id=303277
rdar://165583914

Reviewed by Abrar Rahman Protyasha.

Add basic infrastructure to plumb and evaluate JavaScript-based text extraction filtering rules in a
separate, internal script world as a part of text extraction filtering. Note that there is no change
in behavior yet, since there's no logic to fetch the remotely updatable rule list yet.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRules):

Add a helper method to convert a list of `FilterRuleData` to `FilterRule`, by initializing the URL
regex (if specified).

(WebCore::TextExtraction::filteringWorld):

Set up an internal script world for evaluating text extraction filtering rules.

(WebCore::TextExtraction::applyRules):

Add a helper method that takes a list of text extraction rules, and applies the rules to the `input`
text, each of which consists of a JavaScript snippets to evaluate in the above internal script
world, as well as an optional regex to match against the URL of the given frame.

Each JavaScript snippet takes as arguments the original text, and the container node enclosing the
text; it asynchronously resolves to either a filtered version of the text, or `null` (indicating
that filtering is not necessary).

All snippets are evaluated in parallel, and (to err on the side of the most restrictive rule) the
shortest result is surfaced back through the final completion handler.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

Add a couple new structs, representing data used to initialize filtering rules, and the filtering
rules themselves.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _extractDebugTextWithConfiguration:completionHandler:]):

If the below option flag is set, add another callback to the filtering pipeline that calls out to
the web process to evaluate the extracted text against filtering rules.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:

Add a new option, `_WKTextExtractionFilterRules`, which controls whether or not to enable rule-based
text filtering.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateTextExtractionFilterRules):
(WebKit::WebPageProxy::applyTextExtractionFilter):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add plumbing from the UI -> web process (see above).

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateTextExtractionFilterRules):
(WebKit::WebPage::applyTextExtractionFilter):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/303657@main
